src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Lower.java

```diff
@@ -3518,16 +3518,16 @@ private void visitIterableForeachLoop(JCEnhancedForLoop tree) {
                                               syms.iterableType.tsym);
             if (iterableType.getTypeArguments().nonEmpty())
                 iteratorTarget = types.erasure(iterableType.getTypeArguments().head);
-            Type eType = types.skipTypeVars(tree.expr.type, false);
-            tree.expr.type = types.erasure(eType);
-            if (eType.isCompound())
```

Comment:
This check can needlessly insert casts - it should be irrelevant whether tree.expr.type is a compond type. A cast should be required only when the erased tree.expr.type is not an Iterable.